_posts/2023-09-20-intro-to-the-various-java-based-k8s-oc-tools.adoc

```diff
@@ -0,0 +1,607 @@
+their usages in WildFly related projects.
+---
+
+There are various Java based Kubernetes/OpenShift related tools that
+have different functions, and some of there functions are overlapping.
+Here is the (partial) list of these tools:
+
+* https://dekorate.io/[Dekorate]
+* https://github.com/eclipse/jkube[eclipse/jkube: Build and Deploy java
+applications on Kubernetes]
+* https://github.com/fabric8io/kubernetes-client[fabric8io/kubernetes-client:
+Java client for Kubernetes & OpenShift]
+* https://github.com/fabric8io/docker-maven-plugin[fabric8io/docker-maven-plugin]
+* https://github.com/fabric8io/kubernetes-client[fabric8io/kubernetes-client]
+* https://github.com/GoogleContainerTools/jib[GoogleContainerTools/jib]
+
+All the above tools can be used as Maven plugin, and they can be used
+together if necessary(Which is determined by the requirements). In this
+article I’d like to give a brief introduction to these tools and see
+their usages in some WildFly related projects as examples. Firstly we
+can go through these projects.
+
+== https://dekorate.io/[Dekorate]
+
+From its https://dekorate.io/[website] the project introduction is
+written as:
+
+____
+Dekorate is a one-stop jar to Kubernetes manifest generation that works
+for all jvm languages regardless of your build tool. It makes generating
+Kubernetes manifests as easy as adding a dependency to the classpath.
+Stop wasting time editing xml, json and yml and customize the kubernetes
+manifests as you configure your java application.
+____
+
+In general, the `dekorate` allows users to generate k8s manifest file
+according to the annotations it provided. For example, it provides the
+annotation `@KubernetesApplication` that can trigger the manifest file
+generation. In addition, it also provides annotation `@DockerBuild` that
+can trigger the docker build. Please note that Dekorate does not
+generate `Dockerfile`. It expects to find one in the root of the module
+and it will call the `docker` command to build the image by using the
+installed local docker daemon. There is an article that introduces its
+usagefootnote:[https://developers.redhat.com/blog/2021/03/17/using-dekorate-to-generate-kubernetes-manifests-for-java-applications[Using
+Dekorate to generate Kubernetes manifests for Java applications / Red
+Hat Developer]].
+
+In addition, the Dekorate project provides some supports to various
+frameworks that can better generate the k8s manifest file. For example,
+it contains a framework support for `JAX-RS` based
+projectfootnote:[https://github.com/dekorateio/dekorate/tree/main/frameworks/jaxrs],
+and here is its class diagram:
+
+image:2023-09-20-k8s/dekorate.jpg[image]
+
+The source code of the `ApplicationPathAnnotationProcessor` is like
+this:
+
+[source,java]
+----
+@Description("Detects JAX-RS and servlet annotations and registers the http port.")
+@SupportedAnnotationTypes("javax.ws.rs.ApplicationPath")
```

Comment:
This should likely use the `jakarta` namespace. That said, is dekorate meant to be extended like this? I realize it can be, but should it be?

Reply:
@jamezp  I agree with you. However turrently the implementation is not updated yet: https://github.com/dekorateio/dekorate/blob/main/frameworks/jaxrs/src/main/java/io/dekorate/jaxrs/apt/ApplicationPathAnnotationProcessor.java#L37 So I guess the code for reference should keep same with implementation.

Reply:
Maybe we should leave this off as it won't work in the current version of WildFly.